### PR TITLE
GitHub backend first drop

### PR DESCRIPTION
**Pull request by patrikfirek**
* Create backend through UI (also script added)
* Create repo on backend through UI (also script added)
* Webhooks installed on repo
* Revision created (partly filled)

### CODE REVIEW

File: src/app/CodeRepo/lib/backends/github/github_event_emitter.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
+"use strict";
+
+const log = require("log");
+const Koa = require("/home/patrik/codefarm/src/lib/web/node_modules/koa");
```
Comment (gustfred) [inline src/app/CodeRepo/lib/backends/github/github_event_emitter.js]: a better path maybe?

Comment (patrikfirek) [inline src/app/CodeRepo/lib/backends/github/github_event_emitter.js]: Oops

Comment (patrikfirek) [inline src/app/CodeRepo/lib/backends/github/github_event_emitter.js]: fixed

File: src/app/CodeRepo/lib/backends/github/index.js [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+const moment = require("moment");
+const { AsyncEventEmitter } = require("emitter");
+const GithubEventEmitter = require("./github_event_emitter");
+
+const GITHUB_API_BASE = "https://api.github.com";
+const GITHUB_BASE = "https://github.com";
+
+/*
+To get started with github:
+1. blalbabla
```
Comment (patrikfirek) [inline src/app/CodeRepo/lib/backends/github/index.js]: Need to fill this in with manual steps required

File: src/app/CodeRepo/lib/backends/github/github_event_emitter.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+        this.app.use(async (ctx) => {
+            const body = ctx.request.body;
+            if (body && body.zen) {
+                this.emit("ping", body);
+            } else if (body && body.pull_request) {
+                if (body.action && body.action === "opened") {
+                    this.emit("pull-request-open", body);
+                } else {
+                    console.log("unknown pull-request event received");
```
Comment (christensson) [inline src/app/CodeRepo/lib/backends/github/github_event_emitter.js]: Log using log instead. Level verbose? Replace other console.log as well.

File: src/app/CodeRepo/lib/backends/github/github_event_emitter.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        this.app.close();
+    }
+
+    async start(port) {
+        await this._setupWebServer(port);
+    }
+
+    async dispose() {
+        this.disposed = true;
+        this.teardownWebServer();
```
Comment (christensson) [inline src/app/CodeRepo/lib/backends/github/github_event_emitter.js]: await async function?

File: src/app/CodeRepo/lib/backends/github/index.js [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        this.githubEmitter = new GithubEventEmitter();
+    }
+
+    async start() {
+        // Start event monitoring towards github
+
+        try {
+            await this._startMonitorEventStream();
+        } catch (err) {
+            log.error("Failed to setup local web server: ", err);
```
Comment (christensson) [inline src/app/CodeRepo/lib/backends/github/index.js]: Perhaps a more expressive error message, "Failed to setup GitHub web hook target http server"?

File: src/app/CodeRepo/lib/backends/github/index.js [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        }
+    }
+
+    async validateRepository(/* event, data */) {
+        // TODO: Validate gerrit specific options
+    }
+
+    async _onPing(event) {
+        console.log("ping event received");
```
Comment (christensson) [inline src/app/CodeRepo/lib/backends/github/index.js]: Use log

File: src/app/CodeRepo/lib/backends/github/index.js [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    async _onPullRequestOpen(event) {
+        console.log("pull-request-open received");
+        console.log(event);
```
Comment (christensson) [inline src/app/CodeRepo/lib/backends/github/index.js]: use log

File: src/app/CodeRepo/lib/backends/github/index.js [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+
+        const changeId = event.pull_request.head.sha;
+        const repositoryId = event.repository.name;
+        const repository = await this.Repository.findOne({ _id: repositoryId });
+        if (repository) {
+            const ref = {
+                index: 1,
+                email: null,
```
Comment (christensson) [inline src/app/CodeRepo/lib/backends/github/index.js]: email is important, used by CodeFarm to look-up user

File: src/app/CodeRepo/lib/backends/github/index.js [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+                name: event.pull_request.user.login,
+                submitted: moment.unix(event.pull_request.created_at).utc().format(),
+                comment: event.pull_request.title,
+                change: {
+                    oldrev: null,
+                    newrev: null,
+                    refname: null // Use event.refName all the time instead?
```
Comment (christensson) [inline src/app/CodeRepo/lib/backends/github/index.js]: oldrev, newrev and refname is important.
newrev is the specific commit SHA.
refname is the refname used when pulling the remote changes.

File: src/app/CodeRepo/lib/backends/github/index.js [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    async _startMonitorEventStream() {
+        this.githubEmitter.addListener("ping", this._onPing.bind(this));
+        this.githubEmitter.addListener("pull-request-open", this._onPullRequestOpen.bind(this));
+        await this.githubEmitter.start(this.backend.port);
+    }
+
+    async _createWebHook(repository) {
+        console.log(`github create webhook ${repository._id}`);
```
Comment (christensson) [inline src/app/CodeRepo/lib/backends/github/index.js]: Use log

File: src/app/CodeRepo/lib/backends/github/github_event_emitter.js [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+        console.log("Tearing down web hook server");
+        this.app.close();
+    }
+
+    async start(port) {
+        await this._setupWebServer(port);
+    }
+
+    async dispose() {
+        this.disposed = true;
```
Comment (mattiasrunge) [inline src/app/CodeRepo/lib/backends/github/github_event_emitter.js]: What is the purpose of this disposed flag?

File: src/app/CodeRepo/lib/backends/github/github_event_emitter.js [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+const { AsyncEventEmitter } = require("/home/patrik/codefarm/src/lib/typelib/node_modules/emitter");
+
+class GithubEventEmitter extends AsyncEventEmitter {
+    constructor() {
+        super();
+    }
+
+    async testEventStream() {
+        // Should trigger a ping
+    }
```
Comment (mattiasrunge) [inline src/app/CodeRepo/lib/backends/github/github_event_emitter.js]: Empty method?

File: src/app/CodeRepo/lib/backends/github/index.js [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+const { AsyncEventEmitter } = require("emitter");
+const GithubEventEmitter = require("./github_event_emitter");
+
+const GITHUB_API_BASE = "https://api.github.com";
+const GITHUB_BASE = "https://github.com";
+
+/*
+To get started with github:
+1. blalbabla
+2. blalalalala
```
Comment (mattiasrunge) [inline src/app/CodeRepo/lib/backends/github/index.js]: Useful :)

File: src/app/CodeRepo/lib/backends/github/index.js [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+        const changeId = event.pull_request.head.sha;
+        const repositoryId = event.repository.name;
+        const repository = await this.Repository.findOne({ _id: repositoryId });
+        if (repository) {
+            const ref = {
+                index: 1,
+                email: null,
+                name: event.pull_request.user.login,
+                submitted: moment.unix(event.pull_request.created_at).utc().format(),
+                comment: event.pull_request.title,
```
Comment (mattiasrunge) [inline src/app/CodeRepo/lib/backends/github/index.js]: title indicates to me only the first line, we want the entire message here. But maybe the naming is just confusing?

File: src/app/CodeRepo/lib/backends/github/index.js (hunk 8 above)
Comment (mattiasrunge) [inline src/app/CodeRepo/lib/backends/github/index.js]: Is there no way to get the email?

File: src/app/CodeRepo/lib/backends/github/index.js [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+
+    async _onPing(event) {
+        console.log("ping event received");
+    }
+
+    async _onPullRequestOpen(event) {
+        console.log("pull-request-open received");
+        console.log(event);
+
+        const changeId = event.pull_request.head.sha;
```
Comment (mattiasrunge) [inline src/app/CodeRepo/lib/backends/github/index.js]: Shouldn't this be event.pull_request.id

File: src/app/CodeRepo/lib/backends/github/index.js [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+        const repository = await this.Repository.findOne({ _id: repositoryId });
+        if (repository) {
+            const ref = {
+                index: 1,
+                email: null,
+                name: event.pull_request.user.login,
+                submitted: moment.unix(event.pull_request.created_at).utc().format(),
+                comment: event.pull_request.title,
+                change: {
+                    oldrev: null,
```
Comment (mattiasrunge) [inline src/app/CodeRepo/lib/backends/github/index.js]: Shouldn't this be event.pull_request.base.sha

File: src/app/CodeRepo/lib/backends/github/index.js [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+                    newrev: null,
```
Comment (mattiasrunge) [inline src/app/CodeRepo/lib/backends/github/index.js]: Shouldn't this be event.pull_request.head.sha

File: src/app/CodeRepo/lib/backends/github/index.js (hunk 9 above)
Comment (mattiasrunge) [inline src/app/CodeRepo/lib/backends/github/index.js]: Shouldn't this be event.pull_request.head.ref

File: src/app/CodeRepo/lib/backends/github/index.js [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+
+        const changeId = event.pull_request.head.sha;
+        const repositoryId = event.repository.name;
+        const repository = await this.Repository.findOne({ _id: repositoryId });
+        if (repository) {
+            const ref = {
+                index: 1,
+                email: null,
+                name: event.pull_request.user.login,
+                submitted: moment.unix(event.pull_request.created_at).utc().format(),
```
Comment (mattiasrunge) [inline src/app/CodeRepo/lib/backends/github/index.js]: created_at is when the pull-request is created, what happens when a new commit is added to the pull-request? It seems better with updated_at than created_at to me

File: src/app/CodeRepo/lib/backends/github/index.js [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    async _sendRequest(uri, body, method = "POST") {
+        const auth = Buffer.from(`${this.backend.username}:${this.backend.authToken}`).toString("base64");
+        const options = {
+            method: method,
+            uri: uri,
+            headers: {
+                "User-Agent": "Code Farm",
+                "Authorization": "Basic ".concat(auth)
```
Comment (mattiasrunge) [inline src/app/CodeRepo/lib/backends/github/index.js]: Should use template string `Basic ${auth}`

File: src/app/CodeRepo/lib/backends/github/index.js [hunk 20]
```diff
@@ -0,0 +1,149 @@
...
+            "name": repository._id,
+            "auto_init": true
+        };
+
+        await this._sendRequest(uri, data);
+        await this._createWebHook(repository);
+    }
+
+    async merge(/* repository, revision */) {
+//        log.info(`github merge ${revision._id} in repo ${repository._id}`);
```
Comment (mattiasrunge) [inline src/app/CodeRepo/lib/backends/github/index.js]: Maybe add a throw new Error("Not implemented") and add a TODO?

File: src/app/CodeRepo/lib/backends/github/github_event_emitter.js (hunk 12 above)
Comment (patrikfirek) [inline src/app/CodeRepo/lib/backends/github/github_event_emitter.js]: fixed

File: src/app/CodeRepo/lib/backends/github/github_event_emitter.js (hunk 3 above)
Comment (patrikfirek) [inline src/app/CodeRepo/lib/backends/github/github_event_emitter.js]: fixed

File: src/app/CodeRepo/lib/backends/github/github_event_emitter.js (hunk 11 above)
Comment (patrikfirek) [inline src/app/CodeRepo/lib/backends/github/github_event_emitter.js]: fixed

File: src/app/CodeRepo/lib/backends/github/github_event_emitter.js (hunk 4 above)
Comment (patrikfirek) [inline src/app/CodeRepo/lib/backends/github/github_event_emitter.js]: fixed

File: src/app/CodeRepo/lib/backends/github/index.js (hunk 5 above)
Comment (patrikfirek) [inline src/app/CodeRepo/lib/backends/github/index.js]: fixed

File: src/app/CodeRepo/lib/backends/github/index.js (hunk 6 above)
Comment (patrikfirek) [inline src/app/CodeRepo/lib/backends/github/index.js]: fixed

File: src/app/CodeRepo/lib/backends/github/index.js (hunk 7 above)
Comment (patrikfirek) [inline src/app/CodeRepo/lib/backends/github/index.js]: fixed

File: src/app/CodeRepo/lib/backends/github/index.js (hunk 8 above)
Comment (patrikfirek) [inline src/app/CodeRepo/lib/backends/github/index.js]: now set

Comment (patrikfirek) [inline src/app/CodeRepo/lib/backends/github/index.js]: fixed

File: src/app/CodeRepo/lib/backends/github/index.js (hunk 15 above)
Comment (patrikfirek) [inline src/app/CodeRepo/lib/backends/github/index.js]: fixed

File: src/app/CodeRepo/lib/backends/github/index.js (hunk 16 above)
Comment (patrikfirek) [inline src/app/CodeRepo/lib/backends/github/index.js]: fixed

File: src/app/CodeRepo/lib/backends/github/index.js (hunk 17 above)
Comment (patrikfirek) [inline src/app/CodeRepo/lib/backends/github/index.js]: fixed

File: src/app/CodeRepo/lib/backends/github/index.js (hunk 9 above)
Comment (patrikfirek) [inline src/app/CodeRepo/lib/backends/github/index.js]: pulling from a slave?

Comment (patrikfirek) [inline src/app/CodeRepo/lib/backends/github/index.js]: fixed

File: src/app/CodeRepo/lib/backends/github/index.js (hunk 10 above)
Comment (patrikfirek) [inline src/app/CodeRepo/lib/backends/github/index.js]: fixed

File: src/app/CodeRepo/lib/backends/github/index.js (hunk 19 above)
Comment (patrikfirek) [inline src/app/CodeRepo/lib/backends/github/index.js]: fixed

File: src/app/CodeRepo/lib/backends/github/index.js (hunk 20 above)
Comment (patrikfirek) [inline src/app/CodeRepo/lib/backends/github/index.js]: fixed

File: src/app/CodeRepo/lib/backends/github/index.js [hunk 21]
```diff
@@ -0,0 +1,167 @@
...
+        ServiceMgr.instance.log("debug", event);
+
+        const changeId = event.pull_request.id;
+        const changeSha = event.pull_request.head.sha;
+        const repositoryId = event.repository.name;
+        const email = await this._getCommitAuthor(repositoryId, changeSha);
+        const repository = await this.Repository.findOne({ _id: repositoryId });
+        if (repository) {
+            const ref = {
+                index: 1,
```
Comment (christensson) [inline src/app/CodeRepo/lib/backends/github/index.js]: Need to support more patches. Index needs to be set to correctly...

File: src/app/CodeRepo/lib/backends/github/github_event_emitter.js [hunk 22]
```diff
@@ -10,55 +11,51 @@ class GithubEventEmitter extends AsyncEventEmitter {
...
-            server.once("error", reject);
+            this.server.once("error", reject);
         });
     }
 
     async teardownWebServer() {
-        console.log("Tearing down web hook server");
-        this.app.close();
+        ServiceMgr.instance.log("verbose", "Tearing down web hook server");
+        this.server.close();
```
Comment (christensson) [inline src/app/CodeRepo/lib/backends/github/github_event_emitter.js]: Do we need to close app as well?

File: src/app/CodeRepo/lib/backends/github/index.js [hunk 23]
```diff
@@ -42,43 +42,47 @@ class GithubBackend extends AsyncEventEmitter {
...
 
-        const changeId = event.pull_request.head.sha;
+        const changeId = event.pull_request.id;
+        const changeSha = event.pull_request.head.sha;
         const repositoryId = event.repository.name;
+        const email = await this._getCommitAuthor(repositoryId, changeSha);
         const repository = await this.Repository.findOne({ _id: repositoryId });
         if (repository) {
             const ref = {
                 index: 1,
```
Comment (christensson) [inline src/app/CodeRepo/lib/backends/github/index.js]: Set index correctly for new commits on same PR

File: src/app/CodeRepo/lib/backends/github/index.js [hunk 24]
```diff
@@ -42,43 +42,47 @@ class GithubBackend extends AsyncEventEmitter {
...
-                email: null,
+                email: email,
                 name: event.pull_request.user.login,
                 submitted: moment.unix(event.pull_request.created_at).utc().format(),
                 comment: event.pull_request.title,
                 change: {
-                    oldrev: null,
-                    newrev: null,
-                    refname: null // Use event.refName all the time instead?
+                    oldrev: event.pull_request.base.sha,
```
Comment (christensson) [inline src/app/CodeRepo/lib/backends/github/index.js]: Correct if multiple commits?

File: src/app/CodeRepo/lib/backends/github/github_event_emitter.js (hunk 22 above)
Comment (patrikfirek) [inline src/app/CodeRepo/lib/backends/github/github_event_emitter.js]: there is no function close on app, I couldnt find anything in the examples that indicates the app should be closed

Comment (christensson) [inline src/app/CodeRepo/lib/backends/github/github_event_emitter.js]: OK

**Pull request merged**